### PR TITLE
feature: PresburgerSet subtraction prototype

### LINKED ISSUE #4: Additional set operation: Subtraction + Complement
To go further in the direction of a complete dialect, we need these additional operations. 

Also add canonicalisation passes.

### PR DESCRIPTION
**Pull request by Dinistro**
This contains the first prototype for the Presburger set subtraction. It is taken from libint but does not support the redundant optimisation.

I'm actually not sure if this works, so I decided to already create a draft pull request to enable discussion on the code. I will continue working on the MLIR integration to then be able to test it.

### CODE REVIEW

File: mlir/lib/Dialect/Presburger/Set.cpp [hunk 1]
```diff
@@ -72,6 +73,174 @@ void PresburgerSet::intersectSet(const PresburgerSet &set) {
...
+
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+  };
+
+  for (size_t j = 0; j < S_i.getNumEqualities(); j++) {
+    // The first inequality is positive and the second is negative, of which
+    // we need the complements (strict negative and strict positive).
+    const auto &eq = S_i.getEquality(j);
+    /*if (isMarkedRedundant[2 * j]) {
```
Comment (Dinistro) [inline mlir/lib/Dialect/Presburger/Set.cpp]: Do we need this for correctness or is this just an optimisation? In the later case I suggest that I first try to get a working end to end solution that enables testing.

Comment (superty) [reply mlir/lib/Dialect/Presburger/Set.cpp]: Unfortunately the libint code seems to be incorrect here on master. The NOTs should be inverted here. In fact, I think this heuristic can be omitted until we show that it actually helps noticeably, as it seems to add a good deal of complexity to the code.

I suggest just treating equalities as a pair of inequalities for now, by replacing the body of the loop with:
```cpp
const auto &eq = S_i.getEquality(j);
recurseWithInequalityFromEquality(eq, true, true);
addInequalityFromEquality(eq, false, false);
recurseWithInequalityFromEquality(eq, false, true);
addInequalityFromEquality(eq, true, false);
```

Comment (Dinistro) [reply mlir/lib/Dialect/Presburger/Set.cpp]: Thanks for the feedback. I will adapt this and we can later on continue with this.

File: mlir/lib/Analysis/Presburger/Simplex.cpp [hunk 2]
```diff
@@ -639,17 +649,18 @@ class GBRSimplex {
 };
 
 // Let b_{level}, b_{level + 1}, ... b_n be the current basis.
-// Let F_i(v) = max <v, x - y> where x and y are points in the original polytope
```
Comment (Dinistro) [inline mlir/lib/Analysis/Presburger/Simplex.cpp]: Turns out that my clang-format vim binding doesn't actually change this anymore. Perhaps I accidentally run clang-format with a wrong config at some point. I will fix all the comment changes.

**Comment by Dinistro** [comment]
Closing because these changes are contained in another branch

**Pull request closed without merging**